File: 2024/07/day_07.py
from typing import List, Callable, Dict
from dataclasses import dataclass

OperationDict = Dict[str, Callable[[int, int], int]]

OPERATIONS: OperationDict = {
    '+': lambda x, y: x + y,
    '*': lambda x, y: x * y
}

OPERATIONS_CONCAT: OperationDict = {
    '+': lambda x, y: x + y,
    '*': lambda x, y: x * y,
    '||': lambda x, y: int(str(x) + str(y))
}

@dataclass
class OperatorInput:
    target: int
    values: List[int]

@dataclass
class OperatorOutput:
    operators: List[str]
    target: int
    values: List[int]

class Solution:
    def __init__(self):
        pass
    
    # General Algorithm
    # Recursively iterate on the operators across the values
    # If the value is greater than the target, stop the branch
    # The value is equal to the count, return the set of operators

    def find_operators_recursive(self, target: int, values: List[int], current: int, operator: str, operations: OperationDict) -> List[str]:
        current = operations[operator](current, values[0])
        if current == target and len(values) == 1:
            return [operator]
        if current > target or len(values) == 1:
            return []
        for op in operations.keys():
            result = self.find_operators_recursive(target, values[1:], current, op, operations)
            if result:
                return [operator] + result
        return []
    
        
        
    def find_operators_one(self, input: OperatorInput, operations: OperationDict) -> List[str]:
        for op in operations:
            result = self.find_operators_recursive(input.target, input.values[1:], input.values[0], op, operations)
            if result:
                return result
        return []
    
    def find_operators(self, input: str, operations: OperationDict=OPERATIONS) -> List[OperatorOutput]:
        operators:List[OperatorOutput] = []

        for line in input.splitlines():
            try:
                op_input = OperatorInput(
                    target=int(line.split(":")[0]),
                    values=[int(x) for x in line.split(":")[1].split()]
                )

                ops = self.find_operators_one(op_input, operations)
                operators.append(OperatorOutput(operators=ops, target=op_input.target, values=op_input.values))
            except:
                raise Exception(f"Error processing line: '{line}'")
        return operators
    
    def find_operators_concat(self, input: str) -> List[OperatorOutput]:
        return self.find_operators(input, operations=OPERATIONS_CONCAT)
        
def main(fn: str):
    with open(fn) as f:
        s = Solution()
        data = f.read()
        operators = s.find_operators(data)
        print(sum([x.target for x in operators if x.operators != []]))
        operators = s.find_operators_concat(data)
        print(sum([x.target for x in operators if x.operators != []]))        

if __name__ == "__main__":
    main("input.txt")

File: 2024/07/day_07.jl
using Printf

const OperationsDict = Dict{Char, Function}

OPERATIONS::OperationsDict = Dict(
    '+' => (x, y) -> x + y,
    '*' => (x, y) -> x * y
)

OPERATIONS_CONCAT::OperationsDict = Dict(
    '+' => (x, y) -> x + y,
    '*' => (x, y) -> x * y,
    '|' => (x, y) -> parse(Int64, string(x) * string(y))
)

struct OperatorInput
    target::Int64
    values::Vector{Int64}

    function OperatorInput(target, vals)
        new(target, vals)
    end
end

struct OperatorOutput
    operators::Vector{Char}
    target::Int64
    values::Vector{Int64}

    function OperatorOutput(op, target, vals)
        new(op, target, vals)
    end
end

function find_operators_recursive(target::Int64, values::Vector{Int64}, current::Int64, operator::Char, operations::OperationsDict)::Vector{Char}
    current = operations[operator](current, values[1])
    if current == target && length(values) == 1
        return [operator]
    elseif current > target || length(values) == 1
        return []
    end

    for op in keys(operations)
        result = find_operators_recursive(target, values[2:end], current, op, operations)
        if length(result) > 0
            return [operator; result]
        end
    end

    return []
end

function find_operators_one(input::OperatorInput, operations::OperationsDict)::Vector{Char}
    for op in keys(operations)
        result = find_operators_recursive(input.target, input.values[2:end], input.values[1], op, operations)
        if length(result) > 0
            return result
        end
    end
    return []
end

function find_operators(input::Vector{String}, operations::OperationsDict=OPERATIONS)::Vector{OperatorOutput}
    operators::Vector{OperatorOutput} = []

    for line in input
        parts = split(line, ":")
        target = parse(Int64, strip(parts[1]))
        values = [parse(Int64, x) for x in split(strip(parts[2]), " ")]
        op = find_operators_one(OperatorInput(target, values), operations)
        push!(operators, OperatorOutput(op, target, values))
    end

    return operators
end

function find_operators_concat(input::Vector{String})::Vector{OperatorOutput}
    return find_operators(input, OPERATIONS_CONCAT)
end

function read_data(fn::String)::Vector{String}
    return readlines(fn)
end

if abspath(PROGRAM_FILE) == @__FILE__
    data = read_data("input.txt")

    operators = find_operators(data)
    println(sum([x.target for x in operators if x.operators != []]))

    operators = find_operators_concat(data)
    println(sum([x.target for x in operators if x.operators != []]))

end
